rand_core/src/impls.rs

```diff
@@ -52,17 +52,19 @@ pub fn fill_bytes_via_next<R: RngCore + ?Sized>(rng: &mut R, dest: &mut [u8]) {
     }
 }
 
-trait ToLe: Copy {
+/// Contract: implementing type must be memory-safe to observe as a byte array
+/// (implies no uninitialised padding).
```

Comment:
I'm not sure this means the trait has to be unsafe. How would a safe implementation violating this contract look like?

Reply:
An implementation of `ToLe` for any type that could have padding (e.g. any `repr(Rust)` struct), which may be uninitialized, would be unsound because `fill_via_chunks` is safe and could expose the uninitialized data.

Reply:
@Ralith an unsound implementation of `ToLe` cannot be implemented without `unsafe`, can it? A trait is `unsafe` to indicate that due to the way the trait is used, soundness requires that the impl maintain particular invariants, which I don't see to be the case here. If usage of `ToLe` is sound as long as the implementation is sound, it's a regular trait.

Reply:
An unsound impl is trivial:
```rust
struct ProbablyPadded(u64, u8);
impl ToLe for ProbablyPadded {
    type Bytes = [u8; 0];
    fn to_le_bytes(self) -> { [] }
}
```
The key is that `fill_via_chunks` doesn't use `to_le_bytes` on little-endian targets, it just relies on the unsafe guarantee made by the trait.

Reply:
@Ralith 
Code dependent on `to_le_bytes` does not use any `unsafe`, so it does not make any assumptions about `ToLe` which may cause UB. Your impl will simply panic on `chunk.copy_from_slice(src[i].to_le_bytes().as_ref());`, since chunk size will not be equal to the `src` array.

Reply:
To reiterate, on little-endian targets, `fill_via_chunks` does not invoke `to_le_bytes` ever, instead using `unsafe` to access the `ToLe` implementer's representatoin directly.

Reply:
Thanks @Ralith. There's a subtlety here in that the trait is serving dual roles as a (unsafe) marker trait, and a regular trait.

Reply:
I hadn't put quite so much effort into this trait since it is supposed to change again in the next breaking release, but sure. Perhaps it should be renamed `Observable` with extended documentation?

Reply:
I added a new method to the trait, moving the usage of `unsafe` into the trait.